api: added a validate endpoint

mssola opened the pull request:
Signed-off-by: Miquel Sabaté Solà <msabate@suse.com>

vitoravelino commented:
@mssola I know both endpoints are WIP but the problem I see is that I can't understand what exactly the response means.

Registries:
```json
{
    "messages": null,
    "fields": {
        "reachable": "Error: connection timed out. The given registry is not available!"
    },
    "exists": false,
    "valid": true
}
```

Namespaces:
```json
{
    "messages": null,
    "fields": null,
    "exists": false,
    "valid": true
}
```

What's the difference between `messages` (List of messages for the current status) and `fields` (Detailed hash with the fields)? One is an array and the other is a hash? Do we need both? Hash works better for the client-side.

I think we can move `exists` to the hash, no?

If we could have for namespaces a response like below it would be perfect.

Validating name format (regex) and uniqueness:

```json
// invalid
{
    "messages": {
        "format": "Name can only contain lower case alphanumeric characters, with optional underscores and dashes in the middle",
        "exists": "Name has already been taken"
    },
    "valid": false
}

// valid
{
    "messages": {},
    "valid": true
}
```

If it's too difficult to send the error messages, it could be:

```json
{
    "validations": {
        "format": false,
        "exists": true
    },
    "valid": false
}
```

What do you think?

mssola commented:
> What's the difference between messages (List of messages for the current status) and fields (Detailed hash with the fields)? One is an array and the other is a hash? Do we need both? Hash works better for the client-side.

Yeah, that was clearly and afterthought ... So the idea is that I wanted to reuse the `message` field from the `Messages` entity. Anyways, after that I came up with the `messages` field as an array, because there could be multiple errors, but then when writing the client-side changes, I noticed that we wanted better a hash (e.g. for each field). That's why I added the `fields` hash.

> I think we can move exists to the hash, no?
> 
> If we could have for namespaces a response like below it would be perfect.
> 
> Validating name format (regex) and uniqueness:

Not quite. The `fields` hash is like that because that's what we get from Rails' validation. So, you get a key which is the given attribute (e.g. name) and what's wrong with it (if everything was fine the field is omitted). Moreover, for registries I also added the `reachable` field.

> If it's too difficult to send the error messages, it could be:

I think that a good response that could be implemented in a very easy way is as follows:

```json
// invalid
{
    "messages": {
        "name": "a message that is not unique",
        "exists": true
    },
    "valid": false
}

// valid
{
    "messages": {},
    "valid": true
}
```

What do you think ?

vitoravelino commented:
> Yeah, that was clearly and afterthought ... So the idea is that I wanted to reuse the message field from the Messages entity. Anyways, after that I came up with the messages field as an array, because there could be multiple errors, but then when writing the client-side changes, I noticed that we wanted better a hash (e.g. for each field). That's why I added the fields hash.

Gotcha but we can merge rails validations with our own validations, no? And then we can keep things more consistent and in the same place.

```json
// invalid
{
    "messages": {
        "name": "a message that is not unique",
        "exists": true
    },
    "valid": false
}
```

Mixing boolean with string messages doesn't seem right to me. Also, `name` has more than one kind of validation (format and uniqueness), so how are you gonna display when both validations fail? 

I still believe that the example I provided is more consistent but I have no idea how difficult is to build that kind of response.

mssola commented:
> Mixing boolean with string messages doesn't seem right to me.

I see, that's a good point.

>  Also, name has more than one kind of validation (format and uniqueness), so how are you gonna display when both validations fail?

Well, we can return an array for each field no ? (I think that's what Rails does anyway). So:

```json
// invalid
{
    "messages": {
        "name": ["a message that is not unique"]
    },
    "exists": true,
    "valid": false
}
```

mssola commented:
> I still believe that the example I provided is more consistent but I have no idea how difficult is to build that kind of response.

I agree that it would be better, but I fear we would play catch-up with Rails: we would need to set for each possible validation error a new field. Instead, Rails gives you: this field fails because of [this, this and this]. This is done on purpose for forms, where you want to display below each input "this failed because of this, this and this". This is also a nice format for global error messages.

vitoravelino commented:
> I agree that it would be better, but I fear we would play catch-up with Rails: we would need to set for each possible validation error a new field. Instead, Rails gives you: this field fails because of [this, this and this]. This is done on purpose for forms, where you want to display below each input "this failed because of this, this and this". This is also a nice format for global error messages.

I agree with you. Going with a field for each validation would only make sense if we used boolean as the value.  The front-end would need to know which validation failed to show the specific message. Since we are gonna send the messages themselves, it makes sense to group by field and not validation type. Example updated below:

```json
// invalid
{
    "messages": {
        "name": ["Name can only contain lower case alphanumeric characters, with optional underscores and dashes in the middle", "Name has already been taken"],
    },
    "valid": false
}

// valid
{
    "messages": {},
    "valid": true
}
```

Does that makes sense now?

mssola commented:
> Does that makes sense now?

I think so. Notice that other boolean values will go outside the `messages` hash now (see my example). Other than that, I think we reached an agreement 😄 

I'll update this PR whenever I have the chance.

vitoravelino commented:
Ok. Looking forward to this! 😃

mssola commented:
@vitoravelino feel free to review again. I've also added tests so it should be ready to go. Notice that the registry `/validate` is a bit special since there's only one registry that can be created:

- If no registry exists, then the `valid` boolean field really is the way to go, and everything works as usual (with a `reachable` field in the `messages` telling you all about the reachability of the server).
- I a registry exists, then note that `valid` will return false most of the times (e.g. the name has already been taken), so in this situation you probably want to take a look at the `reachable` field instead (at least for your PR).

mssola commented:
@vitoravelino I need to further update the frontend but I don't know how to use the message of the validator of Vue to the HTML page 😕

vitoravelino commented:
> @vitoravelino I need to further update the frontend but I don't know how to use the message of the validator of Vue to the HTML page 😕

Well, you've changed the return of the validator to string instead of boolean. It has to be a boolean, you can find the documentation at https://monterail.github.io/vuelidate/. What you'll need to do is get the response from the API and set an attribute `errors` on the vue component and do something on the templates based on that. You probably won't understand what I said without some code, but...

Fortunately, there's a similar thing in the registry PR that I worked on. What I did there was to handle the http response on the service (https://github.com/SUSE/Portus/pull/1396/files#diff-4b3854fd3814e22b9cd68969c790ad67R15) because I was relying on http status and then creating the object to pass back to the vue component. But we are already getting the object ready from the API, so I think that doesn't need to be done. In the templates, you have two options:

* Display a message based on booleans (https://github.com/SUSE/Portus/pull/1396/files#diff-28444709d1b74c73b5e97736872005a6R15);
* Display the messages coming from the API (it would be a bit more verbose but still simple);

Since the messages are coming from the API, I would just show them instead of handling the booleans unless there's a really special case for using the booleans.

I did a general comment here that won't help much but I'll make comments inline that will help you understand better what I just said.

vitoravelino commented on app/assets/javascripts/modules/namespaces/components/new-form.js:
```diff
@@ -65,40 +65,28 @@ export default {
           // required already taking care of this
           if (value === '') {
-            return true;
+            return "";
```
Revert this to `true`. We don't need to validate an empty string. That's the responsibility of the required validator that is built-in from the library we are using.

vitoravelino commented on app/assets/javascripts/modules/namespaces/services/namespaces.js:
```diff
@@ -36,9 +36,9 @@ function searchTeam(teamName) {
+  return resource.validate({ name })
+    .then((response) => { return response.data })
     .catch((response) => {
       if (response.status === 404) {
```
Is 404 happening somehow in this endpoint? I don't think so. We can remove this and handle the possible responses properly.

vitoravelino commented on app/assets/javascripts/modules/namespaces/components/new-form.js:
```diff
@@ -65,40 +65,28 @@ export default {
             };
 
-            this.timeout.name = setTimeout(searchName, 1000);
+            this.timeout.name = setTimeout(validate, 1000);
```
Rename `this.timeout.name` to `this.timeout.validate`. Remember to also change on the `data()` method.

vitoravelino commented on app/assets/javascripts/modules/namespaces/components/new-form.js:
```diff
@@ -65,40 +65,28 @@ export default {
-
-                // if exists, invalid
-                resolve(!exists);
+                return "";
```
I would just return `data.valid`.

vitoravelino commented on app/assets/javascripts/modules/namespaces/components/new-form.js:
```diff
@@ -65,40 +65,28 @@ export default {
+              const promise = NamespacesService.validate(value);
+
+              promise.then((data) => {
+                if (data.exists) {
```
Here we could just `set(this, 'errors', data.messages)` and then we can handle properly on the templates.

vitoravelino replied:
Algo need to handle when `null` is returned.

vitoravelino commented on app/views/namespaces/components/_form.html.slim:
```diff
@@ -10,9 +10,7 @@ div ref="form" class="collapse"
           span v-if="!$v.namespace.name.required"
```
Here things are gonna be a little bit different. Something like:

```slim
span.error-messages v-if="errors.name"
  span.error-message v-for="(error, index) in errors.name" :key="index"
    | {{ error }}
```

This will require some styling/positioning fix to break line when multiple errors happen. We can handle that when we reach that state.

mssola commented:
@vitoravelino it's ready to be reviewed now 👍  (I'll squash later)

vitoravelino commented on app/assets/javascripts/modules/namespaces/components/new-form.js:
```diff
@@ -65,40 +68,28 @@ export default {
       name: {
         required,
         format(value) {
```
I think we can rename this to a more generic thing instead of `format`. We are validating everything at once.

vitoravelino commented on app/assets/javascripts/modules/namespaces/components/new-form.js:
```diff
@@ -65,40 +68,26 @@ export default {
-
-                // if exists, invalid
-                resolve(!exists);
+                resolve(data.valid && !data.exists);
```
Forgot to remove `data.exists`.

vitoravelino commented on app/assets/javascripts/modules/namespaces/components/new-form.js:
```diff
@@ -65,40 +68,26 @@ export default {
-                  resolve(true);
+              promise.then((data) => {
+                if (!data.valid) {
+                  set(this.errors, 'name', `Name ${data.messages.name}`);
```
I think just `set(this, 'errors', data.messages)` does the job, no?

vitoravelino commented on app/assets/javascripts/modules/namespaces/components/new-form.js:
```diff
@@ -65,40 +68,26 @@ export default {
-                if (exists === null) {
-                  resolve(true);
+              promise.then((data) => {
+                if (!data.valid) {
```
We need to clean `this.errors` from a previous validation if it's valid. If we don't do that errors messages will still be shown even if data is valid.

We can just simply

```js
set(this, 'errors', data.messages);
```

without any condition here. We already check for errors in the template, so we can do this check if it's empty there.

vitoravelino commented on app/views/namespaces/components/_form.html.slim:
```diff
@@ -9,10 +9,9 @@ div ref="form" class="collapse"
-            | Name can only contain lower case alphanumeric characters, with optional underscores and dashes in the middle
-          span v-if="!$v.namespace.name.available"
-            | Name has already been taken
+          span.error-messages v-if="errors.name"
```
Based on the comment I just left on about setting the the errors messages, I think we can change this condition to `errors.name.length` to avoid the rendering of an empty `span.error-messages`.

vitoravelino commented on app/assets/javascripts/modules/namespaces/components/new-form.js:
```diff
@@ -31,6 +31,9 @@ export default {
         name: null,
```
Rename to `validate` here.

vitoravelino commented on app/assets/javascripts/modules/namespaces/components/new-form.js:
```diff
@@ -28,9 +28,12 @@ export default {
         team: null,
       },
+      errors: {
+        name: '',
```
`[]` instead of `''`, right?
